### DISCUSSION TITLE
Possible migration from custom integration to core integration

### DISCUSSION THREAD

**Discussion by derjoerg** [kingsleyadam/local-abbfreeathome-hass#109] [Q&A]
As - at least from my perspective - the integration is "feature complete" 🥳 (only tests are missing) I wanted to start a discussion how a possible migration from a custom integration to a core integration might work.

I have to admit that I don't know very well the internals of HA so please forgive me, if I ask/write something obvious.

As pointed out by frenck the custom integration should have a different domain as the integration, which should be promoted for core integration.

So are the following high-level steps correct?

1. Fork the HA core repository and copy everything from `abbfreeathome_ci` into a folder under `components` called `abbfreeathome` and also change the domain from `abbfreeathome_ci` to `abbfreeathome`. Create a PR for that (also including a new brands-PR and a documentation-PR)
2. In the custom integration (not linked to the core integration) write "somehow" a detection method, which recognizes that the core integration `abbfreeathome` is available and create a repair issue
3. The repair issue now is responsible for:
3.1. Load ALL data from the custom integration `abbfreeathome_ci` (basic config like host, username and password, but also all modifications like changed entity-names and -ids, icons, ...)
3.2. Delete all data of the custom integration
3.3. Save the loaded data to the new core integration `abbfreeathome`
3.4. Inform the user that all data have been converted and the custom integration can be deleted

Is this correct so far and - much more important - doable?

The reason I'm asking this is that I'm really convinced from this integration and I - till now - don't see any problems in my test environment. So I plan to move this to my prod environment and replace the famous integration from jheling. But
- I have 82 devices with 260 entities
- I've done a lot of modifications in my prod environment regarding naming and so on
- I have a lot of switch-as-x entities
- I have a high number of automations referencing entity-ids
- I have a high number of dashboards referencing entity-ids

I need to modify everything by hand when moving from jheling to this integration and I'm fine with that, even that it will take me several hours.
What I don't want is to make now all these modifications when moving from `freeathome` to `abbfreeathome_ci` and than, once again, when moving from `abbfreeathome_ci` to `abbfreeathome`.

I have no glue for now how a possible detection method would look like and how a possible repair can be programmed, but I'm happy to help and test.

If this is - for whatever reason - not easily possible (what I can fully understand) but as I still want to use this integration in my prod environment, I will do the "dirty" way:
1. Download the custom integration `abbfreeathome_ci`
2. Save it in a folder called `abbfreeathome` within `custom_components`
3. Change the domain in const.py and manifest.json to `abbfreeathome`
4. Track any changes to the custom integration manually
5. Wait for the time the integration becomes a core integration :smiley:
6. Delete the custom component

This should not be offensive. The question was raised and my standpoint was overruled. That's how democracy and especially open-source should work.

**Comment by kingsleyadam** [comment]
I just wanted to bring this up as well. I'm curious if there's any other devices that need to be implemented (that we know of). If not then I'd like to create a 1.0.0 release and provide some cool-down time. Maybe 1-2 weeks to gather any feedback and make any changes for issues that might come-up.

I wanted to implement scenes, specifically the `FID_SPECIAL_SCENE_PANIC` and `FID_SPECIAL_SCENE_ALL_OFF` so I can run automations when those scenes are triggered. But unfortunately those scenes don't seem to come through on the websocket.

To attempt to answer some questions.

> Fork the HA core repository and copy everything from abbfreeathome_ci into a folder under components called abbfreeathome and also change the domain from abbfreeathome_ci to abbfreeathome. Create a PR for that (also including a new brands-PR and a documentation-PR)

I wish it were that easy :). If we attempt to submit this entire integration in a single PR, it'll most certainly be rejected. It's way too much code for a single review. I'm sure there will be feedback. We need to break this down into smaller PR's, starting with the very basics required for the integration and a single platform. 

On the initial PR we'll also need to submit documentation and logos.

I do already have the HA core repo forked. That's actually what I've been working in on all of these changes. After the cool-down period I can try to put together a first PR. Once that's merged in then we move onto the next PR, probably a PR per platform.

> In the custom integration (not linked to the core integration) write "somehow" a detection method, which recognizes that the core integration abbfreeathome is available and create a repair issue

I don't think we'll have that ability to check if abbfreeathome is available in the core. I think both HACS and the core integration should exist and operate independently until all functionality is added to the core integration. My guess is we're not going to get everything added to core in a single release of HA. Once everything is added to the core we can mark this repo as archived and advise those to migrate to the core integration.

> The repair issue now is responsible for:

Have you found any examples of thing being done? This sounds great but I have a feeling it's just goin to be up to the user to delete the custom integration, add the core, and make adjustments as needed. The entity id's should remain the same, any customizations made to the entities would need to be re-done.

> If this is - for whatever reason - not easily possible (what I can fully understand) but as I still want to use this integration in my prod environment, I will do the "dirty" way:

Yep, that's totally doable, especially if we don't actually need many changes in the custom integration going forward (famous last words). Only issue you'll have is the logo won't appear which is purely cosmetic.

**Reply by derjoerg** [reply]
> I just wanted to bring this up as well. I'm curious if there's any other devices that need to be implemented (that we know of). If not then I'd like to create a 1.0.0 release and provide some cool-down time. Maybe 1-2 weeks to gather any feedback and make any changes for issues that might come-up.

100% agree

> I wanted to implement scenes, specifically the `FID_SPECIAL_SCENE_PANIC` and `FID_SPECIAL_SCENE_ALL_OFF` so I can run automations when those scenes are triggered. But unfortunately those scenes don't seem to come through on the websocket.

Regarding the official documentation at https://developer.eu.mybuildings.abb.com/fah_local/concepts scenes are not supported :cry:

> > A SysAP supports configuration of scenes as a way to set multiple datapoints to a configured value by activating a single switch. Common use-cases are “all lights off”, or e.g. “dim living room lights to 50%” etc.
> > A scene in the datamodel of the SysAP is represented as a device with a channel with function ID “4800” (0x4800).
> > However while these special scene devices may be included in the datamodel, they can currently not be controlled directly by the free@home local API: The configuration of scenes is handled completely by the SysAP. The intended way to activate a scene using the free@home local API is to pair the scene with a switch in the SysAP configuration and then activate the switch. 

> I wish it were that easy :). If we attempt to submit this entire integration in a single PR, it'll most certainly be rejected. It's way too much code for a single review. I'm sure there will be feedback. We need to break this down into smaller PR's, starting with the very basics required for the integration and a single platform.

Do you think it is worth to ask this question in the dev-channel on discord?

> On the initial PR we'll also need to submit documentation and logos.

Yep.

> > In the custom integration (not linked to the core integration) write "somehow" a detection method, which recognizes that the core integration abbfreeathome is available and create a repair issue
> 
> I don't think we'll have that ability to check if abbfreeathome is available in the core. I think both HACS and the core integration should exist and operate independently until all functionality is added to the core integration. My guess is we're not going to get everything added to core in a single release of HA. Once everything is added to the core we can mark this repo as archived and advise those to migrate to the core integration.
> 
> > The repair issue now is responsible for:
> 
> Have you found any examples of thing being done? This sounds great but I have a feeling it's just goin to be up to the user to delete the custom integration, add the core, and make adjustments as needed. The entity id's should remain the same, any customizations made to the entities would need to be re-done.

No, but that's how I understood frenck. Or have I missed something in his explanation?

> Yep, that's totally doable, especially if we don't actually need many changes in the custom integration going forward (famous last words). Only issue you'll have is the logo won't appear which is purely cosmetic.

Absolutely

**Reply by derjoerg** [reply]
> I'm curious if there's any other devices that need to be implemented (that we know of).

I also thought about this and had two opinions, which I don't know, which to go for:
1. We leave it as it is (all added devices are tested so far), promote the integration to the community and add additional devices by request (so we also have testers)
2. We check the `freeathome` integration and try map as much of the devices (functionIDs) available there in the `abbfreeathome` library with the draw back, that we don't know for sure if they are working as expected.

**Reply by kingsleyadam** [reply]
> Do you think it is worth to ask this question in the dev-channel on discord?

This has been asked before, here: https://github.com/home-assistant/architecture/issues/408

And I agree with them. A large PR like this makes it impossible to fully review in a reasonable time. I don't want to make it more difficult for them. And I do want to make sure our code meets their standards.

**Reply by kingsleyadam** [reply]
> > I'm curious if there's any other devices that need to be implemented (that we know of).
> 
> 
> 
> I also thought about this and had two opinions, which I don't know, which to go for:
> 
> 1. We leave it as it is (all added devices are tested so far), promote the integration to the community and add additional devices by request (so we also have testers)
> 
> 2. We check the `freeathome` integration and try map as much of the devices (functionIDs) available there in the `abbfreeathome` library with the draw back, that we don't know for sure if they are working as expected.

I think we have to go with option 1. I don't think sifting through the `freeathome` integration is a good use of our time. Mostly because there may be edge cases there that we'll never need to implement here.

If we receive a request for a new device we can always reference that integration to see how it was implemented to ensure we don't miss any possible scenario's. And yea, we should be testing code against physical devices to ensure it operates correctly.

**Reply by kingsleyadam** [reply]
> Regarding the official documentation at https://developer.eu.mybuildings.abb.com/fah_local/concepts scenes are not supported 

This could be interesting, apparently you have to add a virtual switch and associate it with a scene. Whenever I get time I can test this, although it's not high on my list right now.

**Reply by derjoerg** [reply]
> 
> This could be interesting, apparently you have to add a virtual switch and associate it with a scene. Whenever I get time I can test this, although it's not high on my list right now.

As an advanced topic - not so useful, but perhaps for fun - we can implement the `virtualdevice` API-call in the library and provide to HA a service to create specific virtual devices, e.g.
- BinarySensor
- SwitchingActuator
- WindowSensor
- WeatherStation

An interessting usecase e.g. would be
- Create a virtual device of the type "WeatherStation"
- In HA you e.g. have already a weather station setup (e.g. Homematic, or just a weather-provider)
- Let HA populate the different states of the WeatherStation
- You can display the WeatherStation in the F@H app and e.g. a F@H-panel

But sure, for this you don't need to have the "virtualdevice" endpoint done through HA, you can simple do it directly through the swagger-interface of F@H. But, it would be fun :smiley:

**Reply by kingsleyadam** [reply]
> As an advanced topic - not so useful, but perhaps for fun - we can implement the virtualdevice API-call in the library and provide to HA a service to create specific virtual devices, e.g.

I was thinking the same. Maybe for a rainy day. :)

**Reply by derjoerg** [reply]
I also looked at the `notification` endpoint and I got it working through swagger. But the documentation is so ... non-existent and some of the options are totally unclear that I stopped for now.

**Reply by derjoerg** [reply]
> And I agree with them. A large PR like this makes it impossible to fully review in a reasonable time. I don't want to make it more difficult for them. And I do want to make sure our code meets their standards.

As soon as you want to start, don't hesitate to come back to me. I will try to help as much as possible.

**Reply by derjoerg** [reply]
btw: I would go with the `switch` as first entity. It implements both directions (setting and getting) and is just one device.

**Reply by kingsleyadam** [reply]
> As soon as you want to start, don't hesitate to come back to me. I will try to help as much as possible.

I've already got the basic integration stripped down and ready for a PR, but I wanted to write unit tests which are proving to be a bit more difficult than I thought.

**Reply by SolarPower2024** [reply]
I just have discovered this integration. Are you trying to bring it to core soon, or does it still take some time?

I want to try and switch from jheling integration but also am afraid I would need to spend time twice for migration, once for custom integration and later on for core. Which I'd like to avoid and wait for the core...

**Reply by derjoerg** [reply]
> I just have discovered this integration. Are you trying to bring it to core soon, or does it still take some time?

Hi,
I still think that this is the plan from @kingsleyadam - at least I hope so 😜 I recently also thought to bring up this questions again and see what might be fulfilled to start with that. There is also a discussion pending regarding the structure of the underlying module, which might result in some refactoring of the code (based on the result of the discussion it might be meaningful to do this first before working on the core integration). But this all needs to be decided by the maintainer.

> I want to try and switch from jheling integration but also am afraid I would need to spend time twice for migration, once for custom integration and later on for core.

You are absolutely right with your fears with doing double work. You can look at the following discussion (BUT on your own risk):
https://github.com/kingsleyadam/local-abbfreeathome-hass/discussions/118#discussioncomment-11784131

**Reply by kingsleyadam** [reply]
Yea, that is the plan although it may not happen as quickly as I would have liked. There's a decent amount of work to get this into HA core, and it starts with the basics. Likely a single entity type (Switch) and we build on it from there. I'm not sure how many releases will have to be done in HA before we get all entities and functionality into HA core.

We definitely want the Python module to be in it's final form to avoid having to do any sort of refactoring in the HA core project, we're much more flexible as a custom integration for big changes like that.

My free time to work on this fluctuates, so I'm sorry if it seems like I'm not responding daily.

I guess my TLDR is, if you're having issues with the `jheling` integration I wouldn't wait to move to this one. If everything is running smoothly you can wait if you're very patient. :)

**Reply by SolarPower2024** [reply]
Thanks for the feedback.
No rush, don't want to push you.. Was just asking that I'm not starting now migrating and you are in the background already there to start with next release(s) ;-)

Clear to me that it will take time.. Thanks even for spending time to make it happen some time.
